File: benchmarks/NSubstitute.Analyzers.Benchmarks.Source.VisualBasic/DiagnosticsSources/SyncOverAsyncThrowsDiagnosticSource.vb
Imports System
Imports NSubstitute.Analyzers.Benchmarks.Source.VisualBasic.Models
Imports NSubstitute.ExceptionExtensions

Namespace DiagnosticsSources
    Public Class SyncOverAsyncThrowsDiagnosticSource
        Public Sub NS5003_SyncThrowsUsedInAsyncMethod()
            Dim substitute = NSubstitute.Substitute.[For](Of Foo)()
            substitute.TaskReturningAsyncMethod().Throws(New Exception())
            substitute.GenericTaskReturningAsyncMethod().Throws(New Exception())
            substitute.TaskReturningAsyncMethod().Throws(Of Exception)()
            substitute.GenericTaskReturningAsyncMethod().Throws(Of Exception)()
            
            ExceptionExtensions.ExceptionExtensions.Throws(substitute.TaskReturningAsyncMethod(), New Exception())
            ExceptionExtensions.ExceptionExtensions.Throws(substitute.GenericTaskReturningAsyncMethod(), New Exception())
            ExceptionExtensions.ExceptionExtensions.Throws(Of Exception)(substitute.TaskReturningAsyncMethod())
            ExceptionExtensions.ExceptionExtensions.Throws(Of Exception)(substitute.GenericTaskReturningAsyncMethod())
        End Sub
    End Class
End Namespace

File: benchmarks/NSubstitute.Analyzers.Benchmarks.Source.CSharp/DiagnosticsSources/SyncOverAsyncThrowsDiagnosticSource.cs
using System;
using NSubstitute.Analyzers.Benchmarks.Source.CSharp.Models;
using NSubstitute.ExceptionExtensions;

namespace NSubstitute.Analyzers.Benchmarks.Source.CSharp.DiagnosticsSources
{
    public class SyncOverAsyncThrowsDiagnosticSource
    {
        public void NS5003_SyncThrowsUsedInAsyncMethod()
        {
            var substitute = Substitute.For<Foo>();
            substitute.TaskReturningAsyncMethod().Throws(new Exception());
            substitute.GenericTaskReturningAsyncMethod().Throws(new Exception());
            substitute.TaskReturningAsyncMethod().Throws<Exception>();
            substitute.GenericTaskReturningAsyncMethod().Throws<Exception>();

            ExceptionExtensions.ExceptionExtensions.Throws(substitute.TaskReturningAsyncMethod(), new Exception());
            ExceptionExtensions.ExceptionExtensions.Throws(substitute.GenericTaskReturningAsyncMethod(), new Exception());
            ExceptionExtensions.ExceptionExtensions.Throws<Exception>(substitute.TaskReturningAsyncMethod());
            ExceptionExtensions.ExceptionExtensions.Throws<Exception>(substitute.GenericTaskReturningAsyncMethod());
        }
    }
}